cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringControlChannelRequest.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.tethering;
+
+import com.google.gson.Gson;
+import io.cdap.cdap.proto.Notification;
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Sent from TetheringAgentService to TetheringServerHandler.
+ * Contains last TetheringControlMessage received and list of program status update notifications.
+ */
+public class TetheringControlChannelRequest {
+  private static final Gson GSON = new Gson();
+
+  private final String lastControlMessageId;
+  private final String notificationListString;
```

Comment:
this should be a List<Notification> instead of using gson in this class to serialize it. When you use gson to serialize/deserialize the entire TetheringControlChannelRequest, it will do the right thing.

Reply:
fixed